lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -367,6 +384,115 @@ static LogicalResult verifyCircuitOp(CircuitOp circuit) {
       }
     }
   }
+  leafInnerSyms.sort();
+  instanceSyms.sort();
+
+  // Verify the NonLocalAnchor.
+  // 1. Iterate over the namepath.
+  // 2. The namepath should be a valid instance path, specified either on a
+  // module or a declaration inside a module.
+  // 3. Each element in the namepath is an InnerRefAttr except possibly the
+  // last element.
+  // 4. Make sure that the InnerRefAttr is legal, by verifying the module name
+  // and the corresponding inner_sym on the instance.
+  // 5. Make sure that the instance path is legal, by verifying the sequence of
+  // instance and the expected module occurs as the next element in the path.
+  // 6. The last element of the namepath, can be an InnerRefAttr on either a
+  // module port or a declaration inside the module.
+  // 7. The last element of the namepath can also be a module symbol.
+  auto hasNonLocal = [&](AnnotationSet &annos, StringRef nlaName) {
+    bool instFound = false;
+    for (auto anno : annos) {
+      if (auto nlaRef = anno.getMember("circt.nonlocal")) {
+        instFound = nlaRef.cast<FlatSymbolRefAttr>().getAttr() == nlaName;
+        if (instFound)
+          break;
+      }
+    }
+    return instFound;
+  };
+  for (NonLocalAnchor nla : nlaList) {
+    auto namepath = nla.namepath();
+    StringAttr nlaName = nla.sym_nameAttr();
+    if (namepath.size() <= 1) {
+      nla.emitOpError()
+          << "the instance path cannot be empty/single element, it "
+             "must specify an instance path.";
+      return failure();
+    }
+    StringAttr expectedModuleName = {};
+    for (unsigned i = 0, s = namepath.size() - 1; i < s; ++i) {
+      hw::InnerRefAttr innerRef = namepath[i].dyn_cast<hw::InnerRefAttr>();
+      if (!innerRef) {
+        nla.emitOpError()
+            << "the instance path can only contain inner sym reference"
+            << ", only the leaf can refer to a module symbol";
+        return failure();
+      }
+      if (expectedModuleName && expectedModuleName != innerRef.getModule()) {
+        nla.emitOpError() << "instance path is incorrect. Expected module:"
+                          << expectedModuleName
+                          << " instead found:" << innerRef.getModule();
+        return failure();
+      }
+      if (InstanceOp instOp = dyn_cast_or_null<InstanceOp>(
+              instanceSyms.getOpIfExists(innerRef))) {
```

Comment:
Might be better to stick to the "early break" style of code all the way through for consistency:
```c++
InstanceOp instOp = dyn_cast_or_null<InstanceOp>(instanceSyms.getOpIfExists(innerRef);
if (!instOp)
   return nla.emitOpError() << "xxx";

expectedModuleName = instOp.moduleNameAttr().getAttr();
AnnotationSet annos(instOp);
....
```